### PR TITLE
Add HttpTransportFactory class, make credentials serializable, add tests

### PR DESCRIPTION
**Pull request by mziccard**

### CODE REVIEW

**Comment by mziccard** [comment]
This also updates the `guava` dependency from `guava-jdk5:13.0` to `guava:19.0` (see issue #68).

/cc @garrettjonesgoogle @anthmgoogle

File: appengine/java/com/google/auth/appengine/AppEngineCredentials.java [hunk 1]
```diff
@@ -31,41 +31,48 @@
 
 package com.google.auth.appengine;
 
-import com.google.auth.oauth2.AccessToken;
-import com.google.auth.oauth2.GoogleCredentials;
-import com.google.common.collect.ImmutableList;
 import com.google.appengine.api.appidentity.AppIdentityService;
 import com.google.appengine.api.appidentity.AppIdentityService.GetAccessTokenResult;
 import com.google.appengine.api.appidentity.AppIdentityServiceFactory;
+import com.google.appengine.repackaged.com.google.common.collect.ImmutableSet;
```
Comment (garrettjonesgoogle) [inline appengine/java/com/google/auth/appengine/AppEngineCredentials.java]: Don't use repackaged classes

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java [hunk 2]
```diff
@@ -64,8 +68,8 @@
   private final String privateKeyId;
   private final URI defaultAudience;
 
-  // Allow clock to be overriden by test code
-  Clock clock = Clock.SYSTEM;
+  // Until we expose this to the users it can remain transient and non-serializable
+  transient Clock clock = Clock.SYSTEM;
```
Comment (garrettjonesgoogle) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: Should this have `@VisibleForTesting`?

File: oauth2_http/java/com/google/auth/oauth2/UserCredentials.java [hunk 3]
```diff
@@ -155,4 +166,40 @@ public final String getClientSecret() {
...
+  public int hashCode() {
+    return Objects.hash(super.hashCode(), clientId, clientSecret, refreshToken, tokenServerUri,
+        transportFactoryClassName);
+  }
+
+  @Override
+  public String toString() {
+    return toStringHelper()
+        .add("clientId", clientId)
+        .add("clientSecret", clientSecret)
```
Comment (garrettjonesgoogle) [inline oauth2_http/java/com/google/auth/oauth2/UserCredentials.java]: is the client secret a private key?

**Comment by mziccard** [comment]
Also related to #68 we should decide the java version we want to support adding something like the following to the `pom.xml`:
```xml
      <plugin>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>3.5.1</version>
        <configuration>
          <source>1.7</source>
          <target>1.7</target>
          <encoding>UTF-8</encoding>
          <compilerArgument>-Xlint:unchecked</compilerArgument>
        </configuration>
      </plugin>
```

Right now we are not specifying any java version, which defaults us to 1.5. This is a very old version and prevents from using features like try-with-resources and diamond operator. Any reason we are targeting 1.5? Can we explicitly target 1.7 instead?

File: oauth2_http/java/com/google/auth/oauth2/UserCredentials.java (hunk 3 above)
Comment (mziccard) [inline oauth2_http/java/com/google/auth/oauth2/UserCredentials.java]: You are right we should not output this. Fixed

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java (hunk 2 above)
Comment (mziccard) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: Fixed

File: appengine/java/com/google/auth/appengine/AppEngineCredentials.java (hunk 1 above)
Comment (mziccard) [inline appengine/java/com/google/auth/appengine/AppEngineCredentials.java]: Fixed

**Comment by garrettjonesgoogle** [comment]
The change LGTM. It might be good to have one more reviewer. 

Regarding Java 1.5 vs 1.7 - @anthmgoogle should have more background on that than me.

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java [hunk 4]
```diff
@@ -156,6 +162,134 @@ public void getScopes_nullReturnsEmpty() throws IOException {
...
+        tokenServer);
+    OAuth2Credentials otherCredentials = ServiceAccountCredentials.fromPkcs8(SA_CLIENT_ID,
+        SA_CLIENT_EMAIL, SA_PRIVATE_KEY_PKCS8, SA_PRIVATE_KEY_ID, SCOPES, transportFactory,
+        tokenServer);
+    assertTrue(credentials.equals(otherCredentials));
+    assertTrue(otherCredentials.equals(credentials));
+  }
+
+  @Test
+  public void equals_false() throws IOException {
```
Comment (mziccard) [inline oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java]: I am fine with smaller tests. I am happy to split them, will take care of it before merging.

File: oauth2_http/javatests/com/google/auth/oauth2/BaseSerializationTest.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+
+/**
+ * Base class for serialization tests.
+ */
+public class BaseSerializationTest {
+
+  @SuppressWarnings("unchecked")
+  public <T> T serializeAndDeserialize(T obj) throws IOException, ClassNotFoundException {
```
Comment (mziccard) [inline oauth2_http/javatests/com/google/auth/oauth2/BaseSerializationTest.java]: You mean reusing this code in `AppEngineCredentialsTest`? To do that we need to introduce a dependency among test artifacts. This is completely fine (we do that in `google-cloud-java`), just wanted everybody to be aware of it.

**Comment by mziccard** [comment]
@anthmgoogle Thanks for the comments. Are you fine with using `guava 19.0` instead of `guava-jdk`? Also, can we be explicit in targeting java 7 or are there requirements to support older versions? We could use try-with-resources and diamond operators to make our code look cleaner.

**Comment by anthmgoogle** [comment]
@mziccard regarding Java versions and dependencies: The original plan with dependencies for this library was to be consistent with the google-oauth-java-client and related libraries, which were constrained to older versions of Java because of the need to work on Android. For example, 2 years ago you could not using Java compiler 1.7 for that reason and had to stick to 1.6. I don't know if there is any existing use on Android, or if these constraints still apply. To play it safe I would keep the compiler version at 1.6 unless we confirm that 1.7 works on Android now. 

No concerns I'm aware of on the Guava version.

**Comment by lesv** [comment]
https://developer.android.com/guide/platform/j8-jack.html
I'm not recommending a change, just providing data.

**Comment by mziccard** [comment]
From @lesv's link it seems that java7 should be safe to use. Also doing some digging in the library I found out it already uses classes introduced only since 1.7, see [StandardCharsets usage](https://github.com/google/google-auth-library-java/blob/master/oauth2_http/java/com/google/auth/oauth2/ClientId.java#L113), for instance.

**Comment by anthmgoogle** [comment]
Cool. No concerns with these dependencies then.

It sounds like we need to tread more carefully with Java 8.

**Comment by mziccard** [comment]
I added 4 commits that fix all the comments. PTAL.

**Comment by mziccard** [comment]
Can you guys merge this PR? I still don't have write access here. Write access would also allow me to setup Travis. Do you know who has the rights to add users to this repo?

**Comment by anthmgoogle** [comment]
I am one of the admins. I have given you admin access.

**Comment by mziccard** [comment]
Thank you @anthmgoogle

**Pull request merged**